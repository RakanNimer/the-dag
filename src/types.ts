export type GraphNode = {
  id: string | number | symbol;
  nodeID: any;
  nodeData: any;
  possibleTargets: string[];
  possibleSources: string[];
};

export type GraphEdge = {
  edgeID: string;
  edgeData: any;
  sourceID: string;
  targetID: string;
};

export type GraphState = {
  nodes: {
    [nodeID: string]: GraphNode;
  };
  edges: {
    [edgeID: string]: GraphEdge;
  };
};

export type GraphStateMutator = (state: GraphState) => void;

export type GraphStateManipulators = {
  state: GraphState;
  getState: () => GraphState;
  setState: (m: GraphStateMutator) => void;
};

export type PartialNode = {
  nodeID: string;
  data?: {};
  nodeData?: {};
};

export type PartialEdge = {
  source: GraphNode | any;
  target: GraphNode | any;
  edgeData?: any;
};

export type TraverseArgs = {
  startingNodeID: GraphNode["id"];
  visitNode?: (node: GraphNode) => void;
};
export type ImportGraphArg = {
  nodes: any[];
  edges: any[];
  nodeIDGenerator: (node: any) => string;
  edgeSourceIDGenerator: (edge: any) => string;
  edgeTargetIDGenerator: (edge: any) => string;
};
