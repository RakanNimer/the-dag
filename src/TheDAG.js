import getStateManipulators from './core/private/_getStateManipulators';
import { publicMethods } from './core/public/index';

import addEdges from './core/public/addEdges';
import addNodes from './core/public/addNodes';
import isAcyclic from './core/public/isAcyclic';

/**
 * 
 * 
 * @class TheDAG
 * @example 
 * let a = new TheDAG()
 */
class TheDAG {
  constructor(stateManipulators) {
    this.stateManipulators = getStateManipulators(stateManipulators);
  }

  nodeExists(node) {
    const nodeID = node.nodeID ? node.nodeID : node;
    return publicMethods.nodeExists(nodeID, this.stateManipulators);
  }

  /**
    * @param {nodes} nodes
    * @param {Node} nodes.node { nodeID, nodeData }
    * @param {string} nodes.node.nodeID
    * @param {any} nodes.node.nodeData
    * @returns {Void}
    * @memberof TheDAG
   */
  addNodes(nodes) {
    return addNodes(nodes, this.stateManipulators);
  }
  /**
   * Adds Edges to graph.
   * @param {Object[]} Edges
   * @param {Object} Edges.PartialEdge { source, target } 
   * @param {Node | string} PartialEdge.source - Node { nodeID, ...otherStuff} or nodeID string
   * @param {Node | string} PartialEdge.target - Node { nodeID, ...otherStuff} or nodeID string
   * @returns {Void} 
   * @memberof TheDAG
   */
  addEdges(edges) {
    return addEdges(edges, this.stateManipulators);
  }

  /**
   * Gets Edge ID from partial edge information.
   * @param {Object} PartialEdge { source, target } 
   * @param {Node | string} PartialEdge.source - Node { nodeID, ...otherStuff} or nodeID string
   * @param {Node | string} PartialEdge.target - Node { nodeID, ...otherStuff} or nodeID string
   * @returns {string} edgeID
   * @memberof TheDAG
   */
  getEdgeID({ source, target }) {
    return publicMethods.getEdgeID({ source, target }, this.stateManipulators);
  }

  /**
    * Get all nodes in graph
    * @returns {Nodes} graph.nodes
    * @returns {Node} graph.nodes[nodeID] = { nodeID, nodeData }
    * @memberof TheDAG
  */
  getNodes() {
    return publicMethods.getNodes(this.stateManipulators);
  }

  /**
    * Gets all edges in graph
    * @returns {Edges} graph.edges
    * @returns {Edge} graph.edges[edgeID] = { edgeID, nodeData, sourceID, targetID }
    * @memberof TheDAG
   */
  getEdges() {
    return publicMethods.getEdges(this.stateManipulators);
  }

  /**
   * Gets Node by nodeID.
   * 
   * @param {Object} {nodeID} - Any object that has nodeID in it.
   * @param {string} {nodeID}.nodeID - nodeID
   * @returns {Node} { nodeID, nodeData }
   * @memberof TheDAG
   */
  getNode({ nodeID }) {
    return publicMethods.getNode({ nodeID }, this.stateManipulators);
  }
  /**
   * Gets edge by source and target nodes or nodeIDs
   * 
   * @param {Object} VersatileEdge { source, target } 
   * @param {Node | string} VersatileEdge.source - Node { nodeID, ...otherStuff} or nodeID string
   * @param {Node | string} VersatileEdge.target - Node { nodeID, ...otherStuff} or nodeID string
   * @returns {Edge} { edgeID, nodeData, sourceID, targetID }
   * @memberof TheDAG
   */
  getEdge({ source, target }) {
    return publicMethods.getEdge({ source, target }, this.stateManipulators);
  }

  /**
   * 
   * @returns {Object} graph
   * @returns {Nodes} graph.nodes
   * @returns {Node} graph.nodes[nodeID] = { nodeID, nodeData }
   * @returns {Edges} graph.edges
   * @returns {Edge} graph.edges[edgeID] = { edgeID, nodeData, sourceID, targetID }
   * @memberof TheDAG
   */
  toJS() {
    const { nodes, edges } = this.stateManipulators.getState();
    return {
      nodes,
      edges
    };
  }

  /**
   * 
   * Traverses the graph breadth first returning all nodes resulting from the traversal in insertion order. Might result in poor performance on large graphs as it stores it all in memory before returning it.
   * @param {Object}    input 
   * @param {string}    input.startingNodeID - ID of the node from which to start traversing
   * @param {function}  input.visitNode - function called everytime a new node is traversed
   * @memberof TheDAG
   * @yields {Object} generatorOutput
   * @yields {boolean} generatorOutput.done - Reached sink, done traversing.
   * @yields {Object} generatorOutput.value - Node.
    * @yields {string} generatorOutput.value.nodeID - Traversed node ID.
    * @yields {string} generatorOutput.value.nodeData - Traversed node data.
   */

  traverseBreadthFirst({ startingNodeID, visitNode }) {
    return publicMethods.traverseBreadthFirst(
      { startingNodeID, visitNode },
      this.stateManipulators
    );
  }

  /**
   * 
   * Traverses the graph breadth first yielding nodes as it crawls it. Good for large graphs as it doesnt store it all in memory
   * @param {Object}    input 
   * @param {string}    input.startingNodeID - ID of the node from which to start traversing
   * @param {function}  input.visitNode - function called everytime a new node is traversed
   * @memberof TheDAG
   * @yields {Object} generatorOutput
   * @yields {boolean} generatorOutput.done - Reached sink, done traversing.
   * @yields {Object} generatorOutput.value - Node.
    * @yields {string} generatorOutput.value.nodeID - Traversed node ID.
    * @yields {string} generatorOutput.value.nodeData - Traversed node data.
   */
  *traverseBreadthFirstGenerator({ startingNodeID, visitNode = () => null }) {
    yield* publicMethods.traverseBreadthFirstGenerator(
      {
        startingNodeID,
        visitNode
      },
      this.stateManipulators
    );
  }

  /**
   * 
   * Traverses the graph depth first yielding nodes as it crawls it. Good for large graphs as it doesnt store it all in memory
   * @param {Object}    input 
   * @param {string}    input.startingNodeID - ID of the node from which to start traversing
   * @param {function}  input.visitNode - function called everytime a new node is traversed
   * @memberof TheDAG
   * @yields {Object} generatorOutput
   * @yields {boolean} generatorOutput.done - Reached sink, done traversing.
   * @yields {Object} generatorOutput.value - Node.
    * @yields {string} generatorOutput.value.nodeID - Traversed node ID.
    * @yields {string} generatorOutput.value.nodeData - Traversed node data.
   */
  *traverseDepthFirstGenerator({ startingNodeID, visitNode }) {
    yield* publicMethods.traverseDepthFirstGenerator(
      {
        startingNodeID,
        visitNode
      },
      this.stateManipulators
    );
  }

  /**
   * 
   * 
   * @param {Object} input - sourceNodeID and targetNodeID
   * @param {string} input.sourceNodeID - ID of the source node
   * @param {string} input.targetNodeID - ID of the target node
   * @returns {number} numberOfHops - distance between nodes or number of hops required to go from source to target
   * @memberof TheDAG
   */
  getDistanceTo({ sourceNodeID, targetNodeID }) {
    return publicMethods.getDistanceTo(
      { sourceNodeID, targetNodeID },
      this.stateManipulators
    );
  }

  /**
   * 
   * 
   * @param {Object} input - sourceNodeID and hops
   * @param {string} input.sourceNodeID - ID of the source node
   * @param {string} input.hops - Distance or number of hops from source node
   * @returns {Object[]} nodes - All node at distance to source = hops
   * @returns {string} nodes[].nodeID
   * @returns {number} nodes[].hops
   * @memberof TheDAG
   */
  getNodesByDistanceTo({ sourceNodeID, hops }) {
    return publicMethods.getNodesByDistanceTo(
      { sourceNodeID, hops },
      this.stateManipulators
    );
  }

  /**
   * @description
   * Checks if the graph in the object state is acyclic. 
   * If it is acyclic : it returns an array of topologically sorted nodes.
   * If it is cyclic : it returns the ID of the node responsible for the cyclicity of the graph.
   * @returns {Object} acyclicityMetadata
   * @returns {boolean} acyclicityMetadata.isAcyclic
   * @returns {Object[]} acyclicityMetadata.topologicallySortedNodeIDs - A topologically sorted array of node IDs
   * @returns {string} acyclicityMetadata.topologicallySortedNodeIDs[] - nodeID
   * @returns {string} acyclicityMetadata.cyclicDependencyID - If the graph is cyclic, returns the node ID responsable for the cycle. This variable is only set if isAcyclic is false
   * @example 
   * const myDAG = new TheDAG();
   * // ... Add some nodes and edges
   * const { isAcyclic, topologicallySortedNodeIDs } = myDAG.isAcyclic()
   * @memberof TheDAG
   */
  isAcyclic() {
    return isAcyclic(this.stateManipulators);
  }

  /**
   * 
   * Clears nodes and edges in state.
   * @returns {Void}
   * @memberof TheDAG
   */
  destroy() {
    return publicMethods.destroy(this.stateManipulators);
  }
}

export default TheDAG;
