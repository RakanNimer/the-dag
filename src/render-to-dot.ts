export const renderToDot = (
  dagGraph,
  { renderNode = node => "[]", renderEdge = edge => "[]" } = {
    renderNode: () => "[]",
    renderEdge: () => "[]"
  }
) => {
  const { nodes, edges } = dagGraph.toJS();
  const dotString = `
  digraph g {
    ${Object.keys(nodes)
      .map(nodeID => nodes[nodeID])
      .map(node => `NODE_${node.nodeID} ${renderNode(node)};`)
      .join(" ")}
    ${Object.keys(edges)
      .map(edgeID => edges[edgeID])
      .map(
        edge =>
          `NODE_${edge.sourceID} -> NODE_${edge.targetID} ${renderEdge(edge)}`
      )}
  }
`;
  return dotString;
};
