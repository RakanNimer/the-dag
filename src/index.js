import TheDAG from './TheDAG';
export default TheDAG;
