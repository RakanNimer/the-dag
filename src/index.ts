import { TheDAG } from "./TheDAG";
export { TheDAG };
export default TheDAG;
