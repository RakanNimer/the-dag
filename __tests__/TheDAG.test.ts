import { TheDAG } from "../src/";

const createTestGraph = (aDAG, type) => {
  switch (type) {
    case "simple": {
      aDAG.addNodes([{ nodeID: 1 }, { nodeID: 2 }]);
      aDAG.addEdges([{ source: { nodeID: 1 }, target: { nodeID: 2 } }]);
      return;
    }
    case "lessSimple": {
      aDAG.addNodes([
        { nodeID: 1, data: { some: "data" } },
        { nodeID: 2, data: { some: "otherData" } },
        { nodeID: 3, data: {} },
        { nodeID: 4, data: {} },
        { nodeID: 5, data: {} },
        { nodeID: 6, data: {} },
        { nodeID: 7, data: {} },
        { nodeID: 8, data: {} }
      ]);

      aDAG.addEdges([
        { source: 1, target: 3, edgeData: { some: "edgeData" } },
        { source: 1, target: 4 },
        { source: 3, target: 5 },
        { source: 3, target: 6 },
        { source: 4, target: 2 },
        { source: 4, target: 7 },
        { source: 5, target: 8 }
      ]);
      return;
    }
    default:
      return;
  }
};

describe("TheDAG", () => {
  test("exports", () => {
    expect(TheDAG).toMatchSnapshot();
  });
  test("is instantiatable", () => {
    // const getStateManipulators = require("../core/private/_getStateManipulators");
    const aDAG = new TheDAG();
    expect(aDAG).toMatchSnapshot();
  });

  test("import graph from any different format", () => {
    const aDAG = new TheDAG();
    const inputGraphWithDifferentFormat = {
      nodes: [1, 2, 3, 4, 5],
      edges: [
        { source: 1, target: 2 },
        { source: 2, target: 3 },
        { source: 2, target: 4 }
      ]
    };
    const graphReducers = {
      nodeIDGenerator: node => node,
      edgeSourceIDGenerator: edge => edge.source,
      edgeTargetIDGenerator: edge => edge.target
    };
    aDAG.importGraph(
      Object.assign({}, inputGraphWithDifferentFormat, graphReducers)
    );
    expect(aDAG.toJS()).toMatchSnapshot(
      "import graph from any different format"
    );
  });

  test("Create simple graph ", () => {
    const aDAG = new TheDAG();
    createTestGraph(aDAG, "simple");
    expect(aDAG.toJS()).toMatchSnapshot();
    expect(aDAG.nodeExists(1)).toBe(true);
    expect(aDAG.nodeExists({ nodeID: 1 })).toBe(true);
  });

  test("Can get edgeID by loose source and target objects", () => {
    const aDAG = new TheDAG();
    createTestGraph(aDAG, "simple");
    expect(aDAG.getEdgeID({ source: 1, target: 2 })).toBe("1_2");
    expect(
      aDAG.getEdgeID({ source: { nodeID: 1 }, target: { nodeID: 2 } })
    ).toBe("1_2");
  });

  test("can destroy graph nodes and edges", () => {
    const aDAG = new TheDAG();
    createTestGraph(aDAG, "simple");
    aDAG.destroy();
    const nodeIDs = Object.keys(aDAG.toJS().nodes);
    expect(nodeIDs.length).toBe(0);
    const edgeIDs = Object.keys(aDAG.toJS().edges);
    expect(edgeIDs.length).toBe(0);
  });

  test("can create less simple graph", () => {
    const aDAG = new TheDAG();
    createTestGraph(aDAG, "lessSimple");
    expect(aDAG.toJS()).toMatchSnapshot();
  });
  test("Get distance or number of hops required to go from one node to another", () => {
    const aDAG = new TheDAG();
    createTestGraph(aDAG, "lessSimple");
    const distanceFromNodeOneToNodeTwo = aDAG.getDistanceTo({
      sourceNodeID: 1,
      targetNodeID: 2
    });
    expect(distanceFromNodeOneToNodeTwo).toBe(2);
  });
  test("Get edge by loose source and target ", () => {
    const aDAG = new TheDAG();
    createTestGraph(aDAG, "lessSimple");
    const edgeFromOneToThree = aDAG.getEdge({
      source: 1,
      target: 3
    });
    expect({ edgeFromOneToThree }).toMatchSnapshot();
    const edgeFromOneToThreeUsingNodes = aDAG.getEdge({
      source: { nodeID: 1, nodeData: {} },
      target: { nodeID: 3, nodeData: {} }
    });
    expect(edgeFromOneToThree).toEqual(edgeFromOneToThreeUsingNodes);
  });
  test("Get all DAG edges", () => {
    const aDAG = new TheDAG();
    createTestGraph(aDAG, "lessSimple");
    const allDAGEdges = aDAG.getEdges();
    expect(allDAGEdges).toMatchSnapshot();
  });
  test("Get all DAG nodes", () => {
    const aDAG = new TheDAG();
    createTestGraph(aDAG, "lessSimple");
    const allDAGNodes = aDAG.getNodes();
    expect(allDAGNodes).toMatchSnapshot("Get all DAG nodes");
  });
  test("Get node by id", () => {
    const aDAG = new TheDAG();
    createTestGraph(aDAG, "lessSimple");
    const nodeOne = aDAG.getNode({
      nodeID: 1
    });
    expect(nodeOne).toMatchSnapshot("Get node by id");
  });
  test("Traverse the graph breadth first", () => {
    const aDAG = new TheDAG();
    createTestGraph(aDAG, "lessSimple");
    const nodeIterator = aDAG.traverseBreadthFirstGenerator({
      startingNodeID: 1
    });
    let currentNode = nodeIterator.next();
    let orderedNodes = [];
    while (!currentNode.done) {
      orderedNodes.push(currentNode.value);
      currentNode = nodeIterator.next();
    }
    const visitNode = jest.fn();
    const syncTraversalResult = aDAG.traverseBreadthFirst({
      startingNodeID: 1,
      visitNode
    });
    expect(syncTraversalResult).toMatchSnapshot();
    expect(visitNode.mock.calls).toMatchSnapshot();
    expect(orderedNodes).toEqual(syncTraversalResult);
  });
  test("Get nodes by relative distance", () => {
    const aDAG = new TheDAG();
    createTestGraph(aDAG, "lessSimple");
    const nodesTwoHopsAway = aDAG.getNodesByDistanceTo({
      sourceNodeID: 1,
      hops: 2
    });
    expect(nodesTwoHopsAway).toMatchSnapshot();
  });
  test("Check for acyclicity and get topologically sorted array", () => {
    const aDAG = new TheDAG();
    createTestGraph(aDAG, "lessSimple");
    const { isAcyclic, topologicallySortedNodeIDs } = aDAG.isAcyclic();
    expect({ isAcyclic, topologicallySortedNodeIDs }).toMatchSnapshot();
  });

  test("Traverse the graph depth first using generators", () => {
    const aDAG = new TheDAG();
    createTestGraph(aDAG, "lessSimple");
    const nodeIterator = aDAG.traverseDepthFirstGenerator({
      startingNodeID: 1
    });
    let currentNode = nodeIterator.next();
    let orderedNodes = [];
    while (!currentNode.done) {
      orderedNodes.push(currentNode.value);
      currentNode = nodeIterator.next();
    }
    expect(orderedNodes).toMatchSnapshot();
  });
  test("Traverse the graph however you want", () => {
    const aDAG = new TheDAG();
    createTestGraph(aDAG, "lessSimple");
    const nodeIterator = aDAG.traverseDynamicPathGenerator({
      startingNodeID: 1
    });
    let currentNode = nodeIterator.next();
    // console.warn(currentNode);
    let nextNode = nodeIterator.next(
      currentNode.value.visitedNode.possibleTargets[0]
    );
    expect(nextNode).toMatchSnapshot();
    nextNode = nodeIterator.next(nextNode.value.visitedNode.possibleTargets[0]);
    expect(nextNode).toMatchSnapshot();
  });
  test("public methods", () => {
    //   /* Traverse the graph however u want */
    //   const TDP = () => {
    //     const nodeIterator = aDAG.traverseDynamicPathGenerator({
    //       startingNodeID: 1
    //     });
    //     let currentNode = nodeIterator.next();
    //     // console.warn(currentNode);
    //     let nextNode = nodeIterator.next(
    //       currentNode.value.visitedNode.possibleTargets[0]
    //     );
    //     expect(nextNode).toMatchSnapshot();
    //     nextNode = nodeIterator.next(
    //       nextNode.value.visitedNode.possibleTargets[0]
    //     );
    //     expect(nextNode).toMatchSnapshot();
    //   };
    //   TDP();
  });
});
