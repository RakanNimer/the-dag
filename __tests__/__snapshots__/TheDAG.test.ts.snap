// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TheDAG Check for acyclicity and get topologically sorted array 1`] = `
Object {
  "isAcyclic": true,
  "topologicallySortedNodeIDs": Array [
    1,
    3,
    4,
    5,
    6,
    2,
    7,
    8,
  ],
}
`;

exports[`TheDAG Create simple graph  1`] = `
Object {
  "edges": Object {
    "1_2": Object {
      "edgeData": Object {},
      "edgeID": "1_2",
      "sourceID": 1,
      "targetID": 2,
    },
  },
  "nodes": Object {
    "1": Object {
      "id": 1,
      "nodeData": undefined,
      "nodeID": 1,
      "possibleSources": Array [],
      "possibleTargets": Array [
        2,
      ],
    },
    "2": Object {
      "id": 2,
      "nodeData": undefined,
      "nodeID": 2,
      "possibleSources": Array [],
      "possibleTargets": Array [],
    },
  },
}
`;

exports[`TheDAG Get all DAG edges 1`] = `
Object {
  "1_3": Object {
    "edgeData": Object {
      "some": "edgeData",
    },
    "edgeID": "1_3",
    "sourceID": 1,
    "targetID": 3,
  },
  "1_4": Object {
    "edgeData": Object {},
    "edgeID": "1_4",
    "sourceID": 1,
    "targetID": 4,
  },
  "3_5": Object {
    "edgeData": Object {},
    "edgeID": "3_5",
    "sourceID": 3,
    "targetID": 5,
  },
  "3_6": Object {
    "edgeData": Object {},
    "edgeID": "3_6",
    "sourceID": 3,
    "targetID": 6,
  },
  "4_2": Object {
    "edgeData": Object {},
    "edgeID": "4_2",
    "sourceID": 4,
    "targetID": 2,
  },
  "4_7": Object {
    "edgeData": Object {},
    "edgeID": "4_7",
    "sourceID": 4,
    "targetID": 7,
  },
  "5_8": Object {
    "edgeData": Object {},
    "edgeID": "5_8",
    "sourceID": 5,
    "targetID": 8,
  },
}
`;

exports[`TheDAG Get all DAG nodes: Get all DAG nodes 1`] = `
Object {
  "1": Object {
    "id": 1,
    "nodeData": Object {
      "some": "data",
    },
    "nodeID": 1,
    "possibleSources": Array [],
    "possibleTargets": Array [
      3,
      4,
    ],
  },
  "2": Object {
    "id": 2,
    "nodeData": Object {
      "some": "otherData",
    },
    "nodeID": 2,
    "possibleSources": Array [],
    "possibleTargets": Array [],
  },
  "3": Object {
    "id": 3,
    "nodeData": Object {},
    "nodeID": 3,
    "possibleSources": Array [],
    "possibleTargets": Array [
      5,
      6,
    ],
  },
  "4": Object {
    "id": 4,
    "nodeData": Object {},
    "nodeID": 4,
    "possibleSources": Array [],
    "possibleTargets": Array [
      2,
      7,
    ],
  },
  "5": Object {
    "id": 5,
    "nodeData": Object {},
    "nodeID": 5,
    "possibleSources": Array [],
    "possibleTargets": Array [
      8,
    ],
  },
  "6": Object {
    "id": 6,
    "nodeData": Object {},
    "nodeID": 6,
    "possibleSources": Array [],
    "possibleTargets": Array [],
  },
  "7": Object {
    "id": 7,
    "nodeData": Object {},
    "nodeID": 7,
    "possibleSources": Array [],
    "possibleTargets": Array [],
  },
  "8": Object {
    "id": 8,
    "nodeData": Object {},
    "nodeID": 8,
    "possibleSources": Array [],
    "possibleTargets": Array [],
  },
}
`;

exports[`TheDAG Get edge by loose source and target  1`] = `
Object {
  "edgeFromOneToThree": Object {
    "edgeData": Object {
      "some": "edgeData",
    },
    "edgeID": "1_3",
    "sourceID": 1,
    "targetID": 3,
  },
}
`;

exports[`TheDAG Get node by id: Get node by id 1`] = `
Object {
  "id": 1,
  "nodeData": Object {
    "some": "data",
  },
  "nodeID": 1,
  "possibleSources": Array [],
  "possibleTargets": Array [
    3,
    4,
  ],
}
`;

exports[`TheDAG Get nodes by relative distance 1`] = `
Array [
  Object {
    "hops": 2,
    "nodeID": 5,
  },
  Object {
    "hops": 2,
    "nodeID": 6,
  },
  Object {
    "hops": 2,
    "nodeID": 2,
  },
  Object {
    "hops": 2,
    "nodeID": 7,
  },
]
`;

exports[`TheDAG Traverse the graph breadth first 1`] = `
Array [
  Object {
    "hops": 0,
    "nodeID": 1,
  },
  Object {
    "hops": 1,
    "nodeID": 3,
  },
  Object {
    "hops": 1,
    "nodeID": 4,
  },
  Object {
    "hops": 2,
    "nodeID": 5,
  },
  Object {
    "hops": 2,
    "nodeID": 6,
  },
  Object {
    "hops": 2,
    "nodeID": 2,
  },
  Object {
    "hops": 2,
    "nodeID": 7,
  },
  Object {
    "hops": 3,
    "nodeID": 8,
  },
]
`;

exports[`TheDAG Traverse the graph breadth first 2`] = `
Array [
  Array [
    Object {
      "hops": 0,
      "nodeID": 1,
    },
  ],
  Array [
    Object {
      "hops": 1,
      "nodeID": 3,
    },
  ],
  Array [
    Object {
      "hops": 1,
      "nodeID": 4,
    },
  ],
  Array [
    Object {
      "hops": 2,
      "nodeID": 5,
    },
  ],
  Array [
    Object {
      "hops": 2,
      "nodeID": 6,
    },
  ],
  Array [
    Object {
      "hops": 2,
      "nodeID": 2,
    },
  ],
  Array [
    Object {
      "hops": 2,
      "nodeID": 7,
    },
  ],
  Array [
    Object {
      "hops": 3,
      "nodeID": 8,
    },
  ],
]
`;

exports[`TheDAG Traverse the graph depth first using generators 1`] = `
Array [
  Object {
    "id": 3,
    "nodeData": Object {},
    "nodeID": 3,
    "possibleSources": Array [],
    "possibleTargets": Array [
      5,
      6,
    ],
  },
  Object {
    "id": 5,
    "nodeData": Object {},
    "nodeID": 5,
    "possibleSources": Array [],
    "possibleTargets": Array [
      8,
    ],
  },
  Object {
    "id": 8,
    "nodeData": Object {},
    "nodeID": 8,
    "possibleSources": Array [],
    "possibleTargets": Array [],
  },
  Object {
    "id": 6,
    "nodeData": Object {},
    "nodeID": 6,
    "possibleSources": Array [],
    "possibleTargets": Array [],
  },
  Object {
    "id": 4,
    "nodeData": Object {},
    "nodeID": 4,
    "possibleSources": Array [],
    "possibleTargets": Array [
      2,
      7,
    ],
  },
  Object {
    "id": 2,
    "nodeData": Object {
      "some": "otherData",
    },
    "nodeID": 2,
    "possibleSources": Array [],
    "possibleTargets": Array [],
  },
  Object {
    "id": 7,
    "nodeData": Object {},
    "nodeID": 7,
    "possibleSources": Array [],
    "possibleTargets": Array [],
  },
]
`;

exports[`TheDAG Traverse the graph however you want 1`] = `
Object {
  "done": false,
  "value": Object {
    "visitedNode": Object {
      "id": 3,
      "nodeData": Object {},
      "nodeID": 3,
      "possibleSources": Array [],
      "possibleTargets": Array [
        5,
        6,
      ],
    },
  },
}
`;

exports[`TheDAG Traverse the graph however you want 2`] = `
Object {
  "done": false,
  "value": Object {
    "visitedNode": Object {
      "id": 5,
      "nodeData": Object {},
      "nodeID": 5,
      "possibleSources": Array [],
      "possibleTargets": Array [
        8,
      ],
    },
  },
}
`;

exports[`TheDAG can create less simple graph 1`] = `
Object {
  "edges": Object {
    "1_3": Object {
      "edgeData": Object {
        "some": "edgeData",
      },
      "edgeID": "1_3",
      "sourceID": 1,
      "targetID": 3,
    },
    "1_4": Object {
      "edgeData": Object {},
      "edgeID": "1_4",
      "sourceID": 1,
      "targetID": 4,
    },
    "3_5": Object {
      "edgeData": Object {},
      "edgeID": "3_5",
      "sourceID": 3,
      "targetID": 5,
    },
    "3_6": Object {
      "edgeData": Object {},
      "edgeID": "3_6",
      "sourceID": 3,
      "targetID": 6,
    },
    "4_2": Object {
      "edgeData": Object {},
      "edgeID": "4_2",
      "sourceID": 4,
      "targetID": 2,
    },
    "4_7": Object {
      "edgeData": Object {},
      "edgeID": "4_7",
      "sourceID": 4,
      "targetID": 7,
    },
    "5_8": Object {
      "edgeData": Object {},
      "edgeID": "5_8",
      "sourceID": 5,
      "targetID": 8,
    },
  },
  "nodes": Object {
    "1": Object {
      "id": 1,
      "nodeData": Object {
        "some": "data",
      },
      "nodeID": 1,
      "possibleSources": Array [],
      "possibleTargets": Array [
        3,
        4,
      ],
    },
    "2": Object {
      "id": 2,
      "nodeData": Object {
        "some": "otherData",
      },
      "nodeID": 2,
      "possibleSources": Array [],
      "possibleTargets": Array [],
    },
    "3": Object {
      "id": 3,
      "nodeData": Object {},
      "nodeID": 3,
      "possibleSources": Array [],
      "possibleTargets": Array [
        5,
        6,
      ],
    },
    "4": Object {
      "id": 4,
      "nodeData": Object {},
      "nodeID": 4,
      "possibleSources": Array [],
      "possibleTargets": Array [
        2,
        7,
      ],
    },
    "5": Object {
      "id": 5,
      "nodeData": Object {},
      "nodeID": 5,
      "possibleSources": Array [],
      "possibleTargets": Array [
        8,
      ],
    },
    "6": Object {
      "id": 6,
      "nodeData": Object {},
      "nodeID": 6,
      "possibleSources": Array [],
      "possibleTargets": Array [],
    },
    "7": Object {
      "id": 7,
      "nodeData": Object {},
      "nodeID": 7,
      "possibleSources": Array [],
      "possibleTargets": Array [],
    },
    "8": Object {
      "id": 8,
      "nodeData": Object {},
      "nodeID": 8,
      "possibleSources": Array [],
      "possibleTargets": Array [],
    },
  },
}
`;

exports[`TheDAG exports 1`] = `[Function]`;

exports[`TheDAG import graph from any different format: import graph from any different format 1`] = `
Object {
  "edges": Object {
    "1_2": Object {
      "edgeData": Object {},
      "edgeID": "1_2",
      "sourceID": 1,
      "targetID": 2,
    },
    "2_3": Object {
      "edgeData": Object {},
      "edgeID": "2_3",
      "sourceID": 2,
      "targetID": 3,
    },
    "2_4": Object {
      "edgeData": Object {},
      "edgeID": "2_4",
      "sourceID": 2,
      "targetID": 4,
    },
  },
  "nodes": Object {
    "1": Object {
      "id": 1,
      "nodeData": 1,
      "nodeID": 1,
      "possibleSources": Array [],
      "possibleTargets": Array [
        2,
      ],
    },
    "2": Object {
      "id": 2,
      "nodeData": 2,
      "nodeID": 2,
      "possibleSources": Array [],
      "possibleTargets": Array [
        3,
        4,
      ],
    },
    "3": Object {
      "id": 3,
      "nodeData": 3,
      "nodeID": 3,
      "possibleSources": Array [],
      "possibleTargets": Array [],
    },
    "4": Object {
      "id": 4,
      "nodeData": 4,
      "nodeID": 4,
      "possibleSources": Array [],
      "possibleTargets": Array [],
    },
    "5": Object {
      "id": 5,
      "nodeData": 5,
      "nodeID": 5,
      "possibleSources": Array [],
      "possibleTargets": Array [],
    },
  },
}
`;

exports[`TheDAG is instantiatable 1`] = `
TheDAG {
  "stateManipulators": Object {
    "getState": [Function],
    "setState": [Function],
    "state": Object {
      "edges": Object {},
      "nodes": Object {},
    },
  },
}
`;
